File: src/leetCodeProblems/DFSUsingRecursion/BinaryTree/BinaryTreeMaxDepth104.go
package main

/**
- LeetCode - https://leetcode.com/problems/minimum-depth-of-binary-tree/
- Time - O(n)
- Space - O(1)
*/

func calculateHeight(node *TreeNode, height int, maxHeight *int) {
	if node == nil {
		return
	}
	if height > *maxHeight {
		*maxHeight = height
	}

	calculateHeight(node.Left, height+1, maxHeight)
	calculateHeight(node.Right, height+1, maxHeight)
}

func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}

	out := 0
	calculateHeight(root, 0, &out)
	return out + 1
}

// func main() {
// 	root := new(TreeNode) //Returns pointer to TreeNode object
// 	root.Val = 4
// 	root.Left = new(TreeNode)
// 	root.Left.Val = 2
// 	root.Left.Left = new(TreeNode)
// 	root.Left.Left.Val = 1
// 	root.Left.Right = new(TreeNode)
// 	root.Left.Right.Val = 3

// 	root.Right = new(TreeNode)
// 	root.Right.Val = 7
// 	root.Right.Left = new(TreeNode)
// 	root.Right.Left.Val = 6
// 	root.Right.Right = new(TreeNode)
// 	root.Right.Right.Val = 9

// 	log.Println(maxDepth(root))
// }

File: src/leetCodeProblems/DFSUsingRecursion/BinaryTree/BinaryTreeMaxDepth104.java
package leetCodeProblems.BinaryTree;

/**
 * LeetCode - https://leetcode.com/problems/maximum--of-binary-tree/
 */
public class BinaryTreeMax104 {

    public int maxUtil(TreeNode node) {
        if (node == null) {
            return 0;
        }

        System.out.println("Current node value -> " + node.val);
        int leftChildHeight = maxUtil(node.left);
        int rightChildHeight = maxUtil(node.right);

        System.out.println("LeftChildHeight => " + leftChildHeight);
        System.out.println("RightChildHeight => " + rightChildHeight);

        int currentNodeHeight = 1 + Math.max(leftChildHeight, rightChildHeight);

        System.out.println("Current node value -> " + node.val);
        System.out.println("currentNodeHeight => " + currentNodeHeight);

        System.out.println("-----");
        return currentNodeHeight;
    }

    public int max(TreeNode root) {
        return maxUtil(root);
    }

    static class TreeNode {

        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int val) {
            this.val = val;
        }
    }

    public static void main(String[] args) {

        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);
        root.right.left = new TreeNode(6);
        root.right.right = new TreeNode(7);

        BinaryTreeMax104 ob = new BinaryTreeMax104();

        int max = ob.max(root);

        System.out.println(max);

    }
}
